File: cpp/leetcode/editor/cn/[543]二叉树的直径.cpp
#include <vector>
#include "util.h"
#include <map>
#include <stack>
#include <string>
#include <utility>
#include <memory>
#include <iterator>
#include <unordered_map>
#include <algorithm>
#include <numeric>

using namespace std;
//leetcode submit region begin(Prohibit modification and deletion)
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private:
    int maxDiameter = 0;
public:
    int diameterOfBinaryTree(TreeNode *root) {
        maxDepth(root);
        return maxDiameter;
    }

private:

    int maxDepth(TreeNode *root) {
        if (root == nullptr) return 0;

        auto leftMax = maxDepth(root->left);
        auto rightMax = maxDepth(root->right);
        auto myDiameter = leftMax + rightMax;
        maxDiameter = max(myDiameter, maxDiameter);
        return 1 + max(leftMax, rightMax);
    }
};
//leetcode submit region end(Prohibit modification and deletion)

File: dart/leetcode/editor/cn/[543]二叉树的直径.dart
import 'dart:math';

import 'util.dart';
//leetcode submit region begin(Prohibit modification and deletion)

/**
 * Definition for a binary tree node.
 * class TreeNode {
 *   int val;
 *   TreeNode? left;
 *   TreeNode? right;
 *   TreeNode([this.val = 0, this.left, this.right]);
 * }
 */
class Solution {
  int maxDiameter = 0;

  int diameterOfBinaryTree(TreeNode? root) {
    maxDepth(root);
    return maxDiameter;
  }

  int maxDepth(TreeNode? root) {
    if (root == null) return 0;
    var leftDepth = maxDepth(root.left);
    var rightDepth = maxDepth(root.right);
    var myDiameter = leftDepth + rightDepth;
    maxDiameter = max(maxDiameter, myDiameter);

    return max(leftDepth, rightDepth) + 1;
  }
}
//leetcode submit region end(Prohibit modification and deletion)
